File: java2csharp/java/Samsung/ProfessionalAudio.FloatingConotrller.WidgetsLibrary/DeviceActionsLayoutData.java
package com.samsung.android.sdk.professionalaudio.widgets;

import java.util.HashMap;
import java.util.Map;

import android.content.Context;
import android.content.pm.PackageManager.NameNotFoundException;
import android.graphics.drawable.Drawable;
import android.util.SparseArray;

import com.samsung.android.sdk.professionalaudio.app.SapaActionInfo;
import com.samsung.android.sdk.professionalaudio.app.SapaApp;
import com.samsung.android.sdk.professionalaudio.app.SapaAppInfo;

class DeviceActionsLayoutData {
    Drawable mInstanceIcon;
    SapaApp mSapaApp;
    String mInstancePackageName;
    SparseArray<SapaActionInfo> mActionList;
    boolean mIsMultiInstanceEnabled;
    Map<String, SapaActionInfo> mActionMap;
    SapaAppInfo mAppInfo;
    int number = -1;

    DeviceActionsLayoutData() {
        mIsMultiInstanceEnabled = false;
    }

    DeviceActionsLayoutData(SapaAppInfo info, Context context, boolean isMultiInstance) throws NameNotFoundException {
        mAppInfo = info;
        this.mInstanceIcon = info.getIcon(context);
        this.mSapaApp = info.getApp();
        this.mInstancePackageName = info.getPackageName();
        this.mActionList = info.getActions();
        if (this.mActionList == null) {
            this.mActionList = new SparseArray<SapaActionInfo>();
        }
        mActionMap = new HashMap<String, SapaActionInfo>();
        loadActionMap();
        this.mIsMultiInstanceEnabled = info.isMultiInstanceEnabled();
        if (this.mIsMultiInstanceEnabled && isMultiInstance) {
            char c = this.mSapaApp.getInstanceId().charAt(this.mSapaApp.getInstanceId().length() - 1);
            number = Character.digit(c, 10);
            if(number > 0) {
                mInstanceIcon = DrawableTool.getDefaultDrawableWithNumber(mInstanceIcon, number, context);
            }
        }
    }

    void loadActionMap() {
        mActionMap.clear();
        for (int i = mActionList.size()-1; i >=0; --i) {
            SapaActionInfo action = mActionList.valueAt(i);
            mActionMap.put(action.getId(), action);
        }
    }

    public void setActions(SparseArray<SapaActionInfo> ActionList) {
        this.mActionList = ActionList;
        loadActionMap();
    }

    @Override
    public boolean equals(Object o) {
        if (o == null) return false;
        if (this == o) return true;
        if (o instanceof DeviceActionsLayoutData) {
            DeviceActionsLayoutData obj = (DeviceActionsLayoutData) o;
            return (mInstancePackageName.equals(obj.mInstancePackageName) && mSapaApp
                    .equals(obj.mSapaApp));
        }
        return false;
    }
}

File: java2csharp/csharp/Samsung/ProfessionalAudio.FloatingConotrller.WidgetsLibrary/DeviceActionsLayoutData.cs
using System.Collections.Generic;

namespace com.samsung.android.sdk.professionalaudio.widgets
{


	using Context = android.content.Context;
	using NameNotFoundException = android.content.pm.PackageManager.NameNotFoundException;
	using Drawable = android.graphics.drawable.Drawable;
	using SparseArray = android.util.SparseArray;

	using SapaActionInfo = com.samsung.android.sdk.professionalaudio.app.SapaActionInfo;
	using SapaApp = com.samsung.android.sdk.professionalaudio.app.SapaApp;
	using SapaAppInfo = com.samsung.android.sdk.professionalaudio.app.SapaAppInfo;

	internal class DeviceActionsLayoutData
	{
		internal Drawable mInstanceIcon;
		internal SapaApp mSapaApp;
		internal string mInstancePackageName;
		internal SparseArray<SapaActionInfo> mActionList;
		internal bool mIsMultiInstanceEnabled;
		internal IDictionary<string, SapaActionInfo> mActionMap;
		internal SapaAppInfo mAppInfo;
		internal int number = -1;

		internal DeviceActionsLayoutData()
		{
			mIsMultiInstanceEnabled = false;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: DeviceActionsLayoutData(com.samsung.android.sdk.professionalaudio.app.SapaAppInfo info, android.content.Context context, boolean isMultiInstance) throws android.content.pm.PackageManager.NameNotFoundException
		internal DeviceActionsLayoutData(SapaAppInfo info, Context context, bool isMultiInstance)
		{
			mAppInfo = info;
			this.mInstanceIcon = info.getIcon(context);
			this.mSapaApp = info.App;
			this.mInstancePackageName = info.PackageName;
			this.mActionList = info.Actions;
			if (this.mActionList == null)
			{
				this.mActionList = new SparseArray<SapaActionInfo>();
			}
			mActionMap = new Dictionary<string, SapaActionInfo>();
			loadActionMap();
			this.mIsMultiInstanceEnabled = info.MultiInstanceEnabled;
			if (this.mIsMultiInstanceEnabled && isMultiInstance)
			{
				char c = this.mSapaApp.InstanceId.charAt(this.mSapaApp.InstanceId.length() - 1);
				number = char.digit(c, 10);
				if (number > 0)
				{
					mInstanceIcon = DrawableTool.getDefaultDrawableWithNumber(mInstanceIcon, number, context);
				}
			}
		}

		internal virtual void loadActionMap()
		{
			mActionMap.Clear();
			for (int i = mActionList.size() - 1; i >= 0; --i)
			{
				SapaActionInfo action = mActionList.valueAt(i);
				mActionMap[action.Id] = action;
			}
		}

		public virtual SparseArray<SapaActionInfo> Actions
		{
			set
			{
				this.mActionList = value;
				loadActionMap();
			}
		}

		public override bool Equals(object o)
		{
			if (o == null)
			{
				return false;
			}
			if (this == o)
			{
				return true;
			}
			if (o is DeviceActionsLayoutData)
			{
				DeviceActionsLayoutData obj = (DeviceActionsLayoutData) o;
				return (mInstancePackageName.Equals(obj.mInstancePackageName) && mSapaApp.Equals(obj.mSapaApp));
			}
			return false;
		}
	}
}